torch/distributed/rpc/api.py

```diff
@@ -172,38 +173,47 @@ def _wait_all():
             del _thread_local_var.future_list
 
 @_require_initialized
-def _all_gather(obj, timeout=UNSET_RPC_TIMEOUT):
+def _all_gather(obj, worker_names=None, timeout=UNSET_RPC_TIMEOUT):
     r"""
     This is similar to torch.distributed.all_gather(), but is using RPC. It
     picks the worker with the smallest name (alphabetic order) as the leader.
     Then all followers send their data ``obj`` to the leader. After the leader
     has received all, it will broadcast the results back to all followers. This
     function blocks until all workers have received the gathered results.
     """
-    assert (
-        _ALL_WORKER_NAMES is not None
-    ), "`_ALL_WORKER_NAMES` is not initialized for `def _all_gather`."
-    leader_name = sorted(_ALL_WORKER_NAMES)[0]
+    if not worker_names:
+        assert (
+            _ALL_WORKER_NAMES is not None
+        ), "`_ALL_WORKER_NAMES` is not initialized for `def _all_gather`."
+        worker_names = _ALL_WORKER_NAMES
+    leader_name = sorted(worker_names)[0]
 
     self_name = _get_current_rpc_agent().get_worker_info().name
 
-    global _all_gather_sequence_id
-    with _all_gather_dict_lock:
-        sequence_id = _all_gather_sequence_id
-        _all_gather_sequence_id += 1
 
     is_leader = leader_name == self_name
     if timeout == UNSET_RPC_TIMEOUT:
         timeout = get_rpc_timeout()
 
+    if worker_names != _ALL_WORKER_NAMES:
+        # create hash of all worker names in group
+        concat_names = "".join(sorted(worker_names)).encode('utf-8')
+        sequence_id = hashlib.sha256(concat_names).hexdigest()
```

Comment:
There is a race condition when multiple processes using the same sequence ID corresponding to the same set of workers, also need to make this a counter, set(workers) -> sequence_id